### PR TITLE
Add TestingExecutionTrendsView

### PR DESCRIPTION
**Pull request by asankov**
This adds `TestingExecutionTrendsView`. 

Currently, the only functionallity is the chart. While you are reviewing that I will work on implementing `Passing Rate Summary`.

### CODE REVIEW

File: tcms/settings/common.py [hunk 1]
```diff
@@ -105,7 +105,8 @@
     (_('TELEMETRY'), [
         (_('Testing'), [
             (_('Breakdown'), reverse_lazy('testing-breakdown')),
-            (_('Status matrix'), reverse_lazy('testing-status-matrix'))
+            (_('Status matrix'), reverse_lazy('testing-status-matrix')),
+            (_('Execution trends'), reverse_lazy('testing-execution-trends'))
```
Comment (atodorov) [inline tcms/settings/common.py]: add a comma at the end so you don't have to edit the previous line the next time.

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js [hunk 2]
```diff
@@ -142,6 +142,7 @@ function drawChart(data, type, selector) {
     c3.generate(chartConfig);
 }
 
+// TODO: remove this in favour of the one in utils.js
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/breakdown.js]: This belongs in a separate commit/PR

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        const chartData = [['pass'], ['fail']];
+        data.data_set.forEach(entry => {
+            categories.push(`#${entry.run_id}`);
+
+            chartData[0].push(entry.count.pass);
+            chartData[1].push(entry.count.other);
+        });
+
+        //TODO: translations
+        const groups = [['other', 'pass']];
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: all of the names must come from the data set from the backend. FE shouldn't know or care about status names

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        $('#chart > svg').remove();
+
+        const c3ChartDefaults = $().c3ChartDefaults();
+        const stackedHorizontalBarChartConfig = c3ChartDefaults.getDefaultStackedBarConfig();
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: chart must be stacked area chart, not a bar chart, see requirements.

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                tick: {
+                    format: showOnlyRoundNumbers
+                }
+            }
+        };
+        stackedHorizontalBarChartConfig.bindto = '#chart';
+        stackedHorizontalBarChartConfig.color = {
+            pattern: [
+                $.pfPaletteColors.red,
+                $.pfPaletteColors.blue,
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: status colors must match the colors used in status matrix telemetry. Either use CSS or refer to their names as listed in the requirements doc.

File: tcms/telemetry/static/telemetry/js/testing/utils.js [hunk 6]
```diff
@@ -22,3 +22,10 @@ function loadInitialProduct(callback = () => {}) {
 function loadInitialTestPlans() {
     jsonRPC('TestPlan.filter', {}, data => updateSelect(data, '#id_test_plan', 'plan_id', 'name'));
 }
+
+function showOnlyRoundNumbers(number) {
+    if (number % 1 === 0) {
+        return number;
+    }
+    return '';
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/utils.js]: I'm not exactly sure what this function does but if the number isn't round (isn't integer) then we show empty string. Can't we instead round the numbers unconditionally ?

File: tcms/telemetry/urls.py [hunk 7]
```diff
@@ -5,5 +5,7 @@
 urlpatterns = [
     url(r'^testing/breakdown/$', views.TestingBreakdownView.as_view(), name='testing-breakdown'),
     url(r'^testing/status-matrix/$', views.TestingStatusMatrixView.as_view(),
-        name='testing-status-matrix')
+        name='testing-status-matrix'),
+    url(r'^testing/execution-trends/$', views.TestingExecutionTrendsView.as_view(),
+        name='testing-execution-trends')
```
Comment (atodorov) [inline tcms/telemetry/urls.py]: add a comma at the end to avoid editting the previos line next time.

File: tcms/telemetry/api.py [hunk 8]
```diff
@@ -115,3 +115,53 @@ def status_matrix(query=None):
...
+        'other': test_executions.count() - pass_count
+    }
+    row = {'run_id': 0}
+    for test_execution in test_executions:
+        is_pass = test_execution.status.name == TestExecutionStatus.PASSED
+
+        if is_pass:
+            all_count['pass'] += 1
+        else:
+            all_count['other'] += 1
```
Comment (atodorov) [inline tcms/telemetry/api.py]: Design your data structure with IDLE, PASS, FAIL, OTHER (generally speaking just use the status name).

Note: we should probably do some refactoring for these statuses as their meaning can vary from time to time and the most recent PRs need to somehow distinguish statuses without actually hard-coding their names.

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js (hunk 2 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/breakdown.js]: I know, this is why I have put a TODO, rather than move it now

Comment (atodorov) [reply tcms/telemetry/static/telemetry/js/testing/breakdown.js]: Your TODO is totally unrelated to the functionality at hand. Needs to be separated.

Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/breakdown.js]: it is, because, before today we had this function at `breakdown.js`. 
however, now we need it at `execution-trends.js`, so I am extracting it to `utils.js`. now, it must be removed from `breakdown.js`, because we can reuse the one from `utils.js`. however, I do not want to do that change now, so I am putting a TODO, so we don't forget to remove the duplicated code (we did not have duplication until this PR)

File: tcms/telemetry/static/telemetry/js/testing/utils.js (hunk 6 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/utils.js]: the idea of this function is to render the scale label on the y axis. 
why do we need this: because if you have a smal amount of test executions, let's say 3, your 3 test executions would span over the whole screen. the y axis will have multiple labels - 1, 1.1, 1.2, 1.3.. all the way to 3. that of course is useless, because you can't have 1.1 test execution. if we just round the number we will have labels - 1, 1, 1, 1, 1, 1, 2, 2, 2, which is ugly. so we just show the whole numbers - 1, 2, 3.

pictures:
how it is now:
Screenshot 2019-07-11 at 15 23 39
how it is without this function:
Screenshot 2019-07-11 at 15 24 46
how it is if we just round the number:
Screenshot 2019-07-11 at 15 25 34

File: tcms/testruns/models.py [hunk 9]
```diff
@@ -281,10 +295,18 @@ def icon(self):
         with override('en'):
             return self._icons.get(self.name)
 
+    def color(self):
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/testruns/models.py]: `color_code` returns CSS class like: `idle`, `pass`, `fail`
`color` returns actual color like `#72767b`, `#92d400`, `#cc0000`

one is applied on the DOM element, the other is passed to the chart

File: tcms/testruns/models.py [hunk 10]
```diff
@@ -257,9 +256,24 @@ class TestExecutionStatus(TCMSActionModel):
         PASSED: 'pass',
     }
 
+    _chart_statuses = {
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/testruns/models.py]: I removed `_chart_statuses` in favour of `_css_classes`.
What do you mean about the duplication?

**Comment by asankov** [comment]
> X axis labels show vertically when viewing all products - this is not readable, must always be shown horizontally

is that ok? it seems to be horizontal
Screenshot 2019-07-13 at 18 11 17


> X axis labels must be prefixed with TR- if these are test runs. If possible clickable.

done. see picture above. I suppose they can be clickable, but I would have to read more about the DataTables API in order to achieve that. Can't this be added later?

> tooltips are missing a TOTAL count of executions

again - should have to do a bit more reading into the DataTables API

>new telemetry not added in MENU

Done.

File: tcms/settings/common.py (hunk 1 above)
Comment (atodorov) [reply tcms/settings/common.py]: ^^^

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        $('#chart > svg').remove();
+
+        const c3ChartDefaults = $().c3ChartDefaults();
+        const config = c3ChartDefaults.getDefaultAreaConfig();
+        config.axis = {
+            x: {
+                categories: data.categories,
+                type: 'category',
+                tick: {
+                    format: runId => `TR-${runId}`
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: label must be horizontal, not vertical

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            }
+        };
+        config.bindto = '#chart';
+        config.color = {
+            pattern: data.colors
+        };
+        config.data = {
+            columns: chartData,
+            type: 'area-step',
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: Chart type is stacked area chart. Stack with area-spline, see:
https://c3js.org/samples/chart_area_stacked.html

File: tcms/testruns/models.py [hunk 13]
```diff
@@ -18,7 +18,6 @@
 from tcms.xmlrpc.serializer import TestRunXMLRPCSerializer
 from tcms.xmlrpc.utils import distinct_filter
 
-
```
Comment (atodorov) [inline tcms/testruns/models.py]: flake8

File: tcms/testruns/models.py [hunk 14]
```diff
@@ -103,8 +102,8 @@ def add_case_run(self, case, status=1, assignee=None,
             _case_text_version = case.history.latest().history_id
 
         _assignee = assignee \
-            or (case.default_tester_id and case.default_tester) \
-            or (self.default_tester_id and self.default_tester)
+                    or (case.default_tester_id and case.default_tester) \
+                    or (self.default_tester_id and self.default_tester)
```
Comment (atodorov) [inline tcms/testruns/models.py]: flake8

Comment (atodorov) [reply tcms/testruns/models.py]: why do you have to change this ?

File: tcms/testruns/models.py (hunk 13 above)
Comment (atodorov) [reply tcms/testruns/models.py]: why do you have to change this ?

**Comment by atodorov** [comment]
> > X axis labels show vertically when viewing all products - this is not readable, must always be shown horizontally
> is that ok? it seems to be horizontal
> Screenshot 2019-07-13 at 18 11 17

Nope, see
exec-trends

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js (hunk 12 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: The blog post is showing something which is rather `area-step` than `area-spline`, this is why I have chosen this one. Anyway, changed it now

File: tcms/testruns/models.py (hunk 13 above)
Comment (asankov) [reply tcms/testruns/models.py]: Did not. My IDE is trying to be clever

File: tcms/testruns/models.py (hunk 14 above)
Comment (asankov) [reply tcms/testruns/models.py]: Again, my IDE is trying to be clever. Will spend some time configuring it after finishing this one

File: tcms/telemetry/static/telemetry/js/testing/execution-trends.js (hunk 11 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: One option is to do this:
```javascript
config.axis = {
            x: {
                tick: {
                    format: runId => `TR-${runId}`,
                    fit: false,
                    multiline: false,
                },
            },
```
basically saying to DataTables that we want the X axis label to be on line line and be separated from the others. this is the result:
Screenshot 2019-07-14 at 12 38 43
however, in order to achieve that DataTables would just omit some of the labels in order to fit the other ones (see TR-1 label is missing on the picture, along with the other odd labels)

Are you ok with something like this? I could not find an option which would just make the whole column wider, just to fit the labels

Comment (atodorov) [reply tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: go for it

Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/execution-trends.js]: That change is present in this PR. you can review it
